### ISSUE TITLE
Out-of-order samples with Prometheus just scraping itself

### ISSUE THREAD

**Issue by DrJosh9000** [prometheus/prometheus#8394]
**What did you do?**

I'm running Prometheus in a Docker container, scraping itself and a few other containers, and I saw a bunch of logs about out-of-order samples after the first GC. I tried to whittle the config to a minimal reproduction and turns out it basically only needs Prometheus scraping itself to reproduce.

I looked through the other bugs and the mailing list and while there are a few threads about diagnosing out-of-order scrapes, I don't (yet) understand why Prometheus can encounter this behaviour scraping itself- without federation, relabeling, etc.

**What did you expect to see?**

Less log output.

**What did you see instead? Under which circumstances?**

From the logs:

```
...snip...
level=info ts=2021-01-21T09:48:27.705Z caller=main.go:918 msg="Completed loading of configuration file" filename=/etc/prometheus/prometheus.yml totalDuration=3.035631ms remote_storage=2.666µs web_handler=523ns query_engine=1.412µs scrape=110.034µs scrape_sd=71.699µs notify=3.405µs notify_sd=1.197µs rules=1.074µs
level=info ts=2021-01-21T12:22:14.424Z caller=compact.go:507 component=tsdb msg="write block" mint=1611220928609 maxt=1611223200000 ulid=01EWJEAAM5YRDW2DM5MAR1GV89 duration=19.170997ms
level=info ts=2021-01-21T12:22:14.425Z caller=head.go:824 component=tsdb msg="Head GC completed" duration=828.161µs
level=info ts=2021-01-21T13:00:12.739Z caller=compact.go:507 component=tsdb msg="write block" mint=1611223206125 maxt=1611230400000 ulid=01EWJGFVHDRND7YEDGDY1YDRY0 duration=21.721113ms
level=info ts=2021-01-21T13:00:12.740Z caller=head.go:824 component=tsdb msg="Head GC completed" duration=899.59µs
level=info ts=2021-01-21T14:59:52.506Z caller=compact.go:507 component=tsdb msg="write block" mint=1611230401793 maxt=1611237600000 ulid=01EWJQAZ151NAAM4NHJ4PVFT4S duration=20.637383ms
level=info ts=2021-01-21T14:59:52.507Z caller=head.go:824 component=tsdb msg="Head GC completed" duration=919.995µs
level=warn ts=2021-01-21T15:07:38.596Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T15:08:53.554Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T15:10:08.483Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T15:10:38.462Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354

...snip...

level=warn ts=2021-01-21T22:34:34.161Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:35:04.137Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:36:19.095Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:36:34.071Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:38:33.982Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:38:48.982Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=warn ts=2021-01-21T22:39:33.938Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
```

**Environment**

* System information:

Host: `Darwin 20.2.0 x86_64`
Container: `Linux 4.19.121-linuxkit x86_64`

* Prometheus version:

```
prometheus, version 2.24.1 (branch: HEAD, revision: e4487274853c587717006eeda8804e597d120340)
  build user:       root@0b5231a0de0f
  build date:       20210120-00:09:36
  go version:       go1.15.6
  platform:         linux/amd64
```
* Alertmanager version:

N/A

* Prometheus configuration file:
```
global:
  scrape_interval: 15s
  evaluation_interval: 15s

scrape_configs:
  - job_name: 'prometheus'
    static_configs:
      - targets: ['localhost:9090']
  - job_name: 'missing'
    static_configs:
      - targets: ['localhost:12345']
```

I was experimenting with 'missing' job, but I think it is irrelevant - my "production" Prometheus container on the same host is reaching all its targets but has similar log output.

* Alertmanager configuration file:

N/A

* Logs:

See above.

**Comment by brian-brazil** [comment]
Most likely time is broken on your machine somehow. I note you're using a Mac, there was something recently about that having broken monotonic time when suspends are involved - if that's it then it's a OS bug we can't do anything about unfortunately.

Is there anything more from debug logs? Do you have a log of what the current time is on that machine?

**Comment by DrJosh9000** [comment]
Thanks, I'll do some investigating and get back to you.

**Comment by DrJosh9000** [comment]
Debug logs: turned it on, I don't see anything unusual, but I'm not sure what to look for among the flood of messages of the form:

```
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series=prometheus_tsdb_wal_writes_failed_total
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series=prometheus_web_federation_errors_total
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series=prometheus_web_federation_warnings_total
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series=promhttp_metric_handler_requests_in_flight
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series="promhttp_metric_handler_requests_total{code=\"200\"}"
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series="promhttp_metric_handler_requests_total{code=\"500\"}"
level=debug ts=2021-01-22T08:58:53.673Z caller=scrape.go:1414 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Out of order sample" series="promhttp_metric_handler_requests_total{code=\"503\"}"
level=warn ts=2021-01-22T08:58:53.673Z caller=scrape.go:1372 component="scrape manager" scrape_pool=prometheus target=http://localhost:9090/metrics msg="Error on ingesting out-of-order samples" num_dropped=354
level=debug ts=2021-01-22T08:58:58.762Z caller=scrape.go:1127 component="scrape manager" scrape_pool=missing
```

System clock: any suggestions for trying to measure time skew on the host or in the containers appreciated. Based on the system.log, `timed` has only called `settimeofday` 13 times in the last day, which to me doesn't sound like enough to cause the almost continuous out-of-order samples, so I'll focus on Docker I guess?

**Comment by brian-brazil** [comment]
Do you have a longer selection of logs? The timestamps on the log messages may help figure out what's going on.

**Comment by beorn7** [comment]
Another possible reason might be that the machine (or VM) your container is running on was suspended and woken up.

In any case, more information would be helpful, or any new hints you have found.

**Comment by DrJosh9000** [comment]
Apologies for the lack of updates, I'm continuing to try to isolate the problem, and migrated the real container over to a Linux box which, unsurprisingly, doesn't have this issue. It hardly seems to happen when I'm looking at the logs from the Docker Desktop console, which I suppose is evidence that the monotonic clock is broken when the machine is not fully active - basically what Brian suggested.

Feel free to close this. If I find some better evidence, I'll let you know.

**Comment by beorn7** [comment]
Thank you. Will close now, but feel free to amend with more evidence.

**Comment by roidelapluie** [comment]
This is probably a duplicate of #8243 fixed by #8601.
